File: 128. Longest Consecutive Sequence/128.cpp
/*
- hashmaps have O(1) find time complexity
- O(n) can mean multiple independent iterations

- time: O(n)
- space: O(n)
*/
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        // true: visited, false: not yet
        unordered_map<int, bool> ele_map;
        int maxlen = 0;
        
        for(int num: nums){
            ele_map[num] = false;
        }

        for(int num: nums){
            if (ele_map[num]){
                continue;
            }

            int templen = 1;
            int smallernum = num - 1, largernum = num + 1;

            while(ele_map.find(smallernum) != ele_map.end()){
                ele_map[smallernum--] = true;
                templen++;
            }

            while(ele_map.find(largernum) != ele_map.end()){
                ele_map[largernum++] = true;
                templen++;
            }

            maxlen = max(maxlen, templen);
        }

        return maxlen;
    }
};

File: 128. Longest Consecutive Sequence/128.py
"""
O(1) searching -> set or dictionary
time: O(n)
space: O(n)
"""
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:

        maxlen = 0 
        nums_set = set(nums)
        for num in nums_set:
            if num - 1 not in nums_set:
                current_len = 1
                current_num = num
                while current_num + 1 in nums_set:
                    current_len += 1
                    current_num += 1
                
                maxlen = max(maxlen, current_len)

        return maxlen

"""
edge: if empty return 0
maxlen
intial -> sort -> keep track of previous element and update maxlen

convert nums to set
for num in nums_set:
    if num - 1 not in nums_set:
        current_len = 1
        current_num = num
        repeatedly search for current_num + 1 and update current_len

    update maxlen
"""
